### PR TITLE
Recursively close / expand trees when holding alt

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
We change the top-down / bottom-up views to collapse / expand
recursively if the user makes a click with holding the ALT key.

Bug: http://b/202129139

### CODE REVIEW

**Comment by dpallotti** [comment]
I need to ask the "why" first. The bug also doesn't say anything. I tried a few tree views in a couple of programs (Windows Explorer, CLion, Visual Studio, even VTune), and none of them expand recursively on ALT+click.

**Comment by vickyliu-go4it** [comment]
> I need to ask the "why" first. The bug also doesn't say anything. I tried a few tree views in a couple of programs (Windows Explorer, CLion, Visual Studio, even VTune), and none of them expand recursively on ALT+click.

Really appreciate for checking the tree view behavior of other programs!

I don't remember where this comes from 🙈  Our sub team discussed it last year and this got a high priority in our hotlist after discussion. I will double check with Florian when he comes back.

File: src/OrbitQt/CallTreeWidget.cpp [hunk 1]
```diff
@@ -351,6 +353,17 @@ static void CollapseRecursively(QTreeView* tree_view, const QModelIndex& index)
...
 }
 
+void CallTreeWidget::OnAltKeyAndMousePressed(const QPoint& point) {
+  QModelIndex index = ui_->callTreeTreeView->indexAt(point);
+  if (!index.isValid()) return;
+
+  if (ui_->callTreeTreeView->isExpanded(index)) {
+    CollapseRecursively(ui_->callTreeTreeView, index);
+  } else {
+    ExpandRecursively(ui_->callTreeTreeView, index);
```
Comment (dpallotti) [inline src/OrbitQt/CallTreeWidget.cpp:363]: After https://github.com/google/orbit/pull/3499 that I just merged, this requires wrapping with `DisableThreadOrFunctionColumnResizeOnRowExpanded();` (before) and `ReEnableThreadOrFunctionColumnResizeOnRowExpanded(); ResizeThreadOrFunctionColumnToShowVisibleDescendants(selected_index);` (after). Have a look at `CallTreeWidget::OnCustomContextMenuRequested`. Not ideal, but see that pull request for an explanation.

File: src/OrbitQt/CopyKeySequenceEnabledTreeView.h [hunk 2]
```diff
@@ -6,6 +6,7 @@
 #define ORBIT_QT_COPYABLE_TREE_VIEW_H_
 
 #include <QKeyEvent>
+#include <QMouseEvent>
 #include <QTreeView>
 
 class CopyKeySequenceEnabledTreeView : public QTreeView {
```
Comment (dpallotti) [inline src/OrbitQt/CopyKeySequenceEnabledTreeView.h:12]: The name of this class was tightly coupled to the one signal it was adding. I knew the time would come when we would add another signal and make the name obsolete. So unless there is a way to do the `altKeyAndMousePressed` with only existing signals, we need to rename the class. 
I'm not super sure about the name though. `AdditionalSignalsTreeView`? `CustomSignalsTreeView`? 
We could also tie it to the fact that it's used for the "call tree"s and call it `CallTreeTreeView` but I think we can keep it generic, there is nothing in this class making it specific for the top-down and bottom-up views.

File: src/OrbitQt/CopyKeySequenceEnabledTreeView.h [hunk 3]
```diff
@@ -22,8 +23,19 @@ class CopyKeySequenceEnabledTreeView : public QTreeView {
     }
   }
 
+  void mousePressEvent(QMouseEvent* event) override {
+    QModelIndex index = indexAt(event->pos());
+    bool alt = event->modifiers() & Qt::AltModifier;
```
Comment (dpallotti) [inline src/OrbitQt/CopyKeySequenceEnabledTreeView.h:28]: I'd suggest `alt_pressed`.

**Review by dpallotti — COMMENTED** [review]
Regarding using ALT+click in the first place, as discussed offline it seems that this comes from macOS. I'd still double check with @reichlfl as this is not standard on Windows (nor Linux; I'm even wondering if this will work on Linux as ALT+click allows you to move an entire window without having to drag from the title bar).

**Comment by reichlfl** [comment]
> Regarding using ALT+click in the first place, as discussed offline it seems that this comes from macOS. I'd still double check with @reichlfl as this is not standard on Windows (nor Linux; I'm even wondering if this will work on Linux as ALT+click allows you to move an entire window without having to drag from the title bar).

As I discussed offline with Vicky: I don't really have an opinion on this. It was Carsten's suggestion, so I'd ask to have him make the final call here. The point about it being potentially conflicting in Linux is valid, so maybe we want to choose a different hotkey - it seems like there is no real standard. I can see this functionality being quite useful though once you're used to it. We can always mention the hotkey in the context menu, can't we?

**Comment by dpallotti** [comment]
> > Regarding using ALT+click in the first place, as discussed offline it seems that this comes from macOS. I'd still double check with @reichlfl as this is not standard on Windows (nor Linux; I'm even wondering if this will work on Linux as ALT+click allows you to move an entire window without having to drag from the title bar).
> 
> As I discussed offline with Vicky: I don't really have an opinion on this. It was Carsten's suggestion, so I'd ask to have him make the final call here. The point about it being potentially conflicting in Linux is valid, so maybe we want to choose a different hotkey - it seems like there is no real standard. I can see this functionality being quite useful though once you're used to it. We can always mention the hotkey in the context menu, can't we?

Another key would be CTRL or SHIFT. But those are used to select multiple rows (respectively, individual or consecutive). The behaviour would be unexpected.

**Comment by vickyliu-go4it** [comment]
> > > Regarding using ALT+click in the first place, as discussed offline it seems that this comes from macOS. I'd still double check with @reichlfl as this is not standard on Windows (nor Linux; I'm even wondering if this will work on Linux as ALT+click allows you to move an entire window without having to drag from the title bar).
> > 
> > 
> > As I discussed offline with Vicky: I don't really have an opinion on this. It was Carsten's suggestion, so I'd ask to have him make the final call here. The point about it being potentially conflicting in Linux is valid, so maybe we want to choose a different hotkey - it seems like there is no real standard. I can see this functionality being quite useful though once you're used to it. We can always mention the hotkey in the context menu, can't we?
> 
> Another key would be CTRL or SHIFT. But those are used to select multiple rows (respectively, individual or consecutive). This is standard. Changing that behaviour would be unexpected and potentially annoying.

Carsten also mentioned that SHIFT and CTRL might be conflicting with selections.    
Also maybe we don't need to restrict to these common modifier keys, we can also pick other keys and mention in the context menu.

**Comment by vickyliu-go4it** [comment]
@reichlfl , @dpallotti  After the offline discussion, we decided that the “ALT+Click” shortcut should be displayed next to the action (i.e., expand / collapse recursively) that matches click behavior.

See the comments in the bug for more details http://b/202129139

**Review by reichlfl — APPROVED** [review]

**Pull request merged**